interface value in map

The following program `sample.go` triggers a panic:

```go
package main

import "fmt"
    
func main() {
    var a = map[string]interface{}{"test": "test"}
    fmt.Println(a["test"])
}
```

Expected result:
```console
$ go run ./sample.go
test
```

Got:
```console
$ yaegi ./sample.go
~/.sample.go:6:33: panic
reflect.Set: value of type string is not assignable to type interp.valueInterface
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc0003a2000, 0x40)
        /home/marc/sdk/go1.14.1/src/runtime/debug/stack.go:24 +0x9d
github.com/containous/yaegi/interp.(*Interpreter).Eval.func1(0xc000299e10)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:320 +0xbc
panic(0xc93360, 0xc000212690)
        /home/marc/sdk/go1.14.1/src/runtime/panic.go:967 +0x166
github.com/containous/yaegi/interp.runCfg.func1(0xc00037e0a0, 0xc000223500)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:113 +0x20c
panic(0xc93360, 0xc000212690)
        /home/marc/sdk/go1.14.1/src/runtime/panic.go:967 +0x166
reflect.Value.assignTo(0xc93360, 0xc000212490, 0x98, 0xdc4944, 0xb, 0xd16a00, 0x0, 0x15, 0xc93360, 0xc000212490)
        /home/marc/sdk/go1.14.1/src/reflect/value.go:2403 +0x426
reflect.Value.Set(0xd16a00, 0xc0002187c0, 0x199, 0xc93360, 0xc000212490, 0x98)
        /home/marc/sdk/go1.14.1/src/reflect/value.go:1532 +0xbd
github.com/containous/yaegi/interp.getIndexMap.func2(0xc00037e0a0, 0xc00021c3c0)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:1059 +0x158
github.com/containous/yaegi/interp.runCfg(0xc000223500, 0xc00037e0a0)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:119 +0x6d
github.com/containous/yaegi/interp.(*Interpreter).run(0xc000202300, 0xc000222600, 0xc00037e000)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:97 +0x239
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc000202300, 0xc000384000, 0x74, 0x0, 0x0, 0x0, 0x0, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:388 +0x41f
main.main()
        /home/marc/go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:94 +0x548
```

Thanks to @juliens for reporting this.

--- a/_test/interface31.go
+++ b/_test/interface31.go
@@ -1,0 +1,11 @@
+package main
+
+import "fmt"
+
+func main() {
+	s := []interface{}{"test", 2}
+	fmt.Println(s[0], s[1])
+}
+
+// Output:
+// test 2

--- a/_test/interface32.go
+++ b/_test/interface32.go
@@ -1,0 +1,11 @@
+package main
+
+import "fmt"
+
+func main() {
+	s := [2]interface{}{"test", 2}
+	fmt.Println(s[0], s[1])
+}
+
+// Output:
+// test 2

--- a/_test/interface33.go
+++ b/_test/interface33.go
@@ -1,0 +1,11 @@
+package main
+
+import "fmt"
+
+func main() {
+	var a = map[string]interface{}{"test": "test"}
+	fmt.Println(a["test"])
+}
+
+// Output:
+// test

--- a/_test/interface34.go
+++ b/_test/interface34.go
@@ -1,0 +1,11 @@
+package main
+
+import "fmt"
+
+func main() {
+	s := [2]interface{}{1: "test", 0: 2}
+	fmt.Println(s[0], s[1])
+}
+
+// Output:
+// 2 test

--- a/_test/interface35.go
+++ b/_test/interface35.go
@@ -1,0 +1,15 @@
+package main
+
+import "fmt"
+
+type T struct {
+	I interface{}
+}
+
+func main() {
+	t := T{"test"}
+	fmt.Println(t)
+}
+
+// Output:
+// test

--- a/_test/interface35.go
+++ b/_test/interface35.go
@@ -12,4 +12,4 @@ func main() {
 }
 
 // Output:
-// test
+// {test}

--- a/interp/run.go
+++ b/interp/run.go
@@ -1580,11 +1580,19 @@ func arrayLit(n *node) {
 	for i, c := range child {
 		if c.kind == keyValueExpr {
 			convertLiteralValue(c.child[1], rtype)
-			values[i] = genValue(c.child[1])
+			if n.typ.val.cat == interfaceT {
+				values[i] = genValueInterface(c.child[1])
+			} else {
+				values[i] = genValue(c.child[1])
+			}
 			index[i] = int(vInt(c.child[0].rval))
 		} else {
 			convertLiteralValue(c, rtype)
-			values[i] = genValue(c)
+			if n.typ.val.cat == interfaceT {
+				values[i] = genValueInterface(c)
+			} else {
+				values[i] = genValue(c)
+			}
 			index[i] = prev
 		}
 		prev = index[i] + 1
@@ -1593,14 +1601,14 @@ func arrayLit(n *node) {
 		}
 	}
 
-	var a reflect.Value
-	if n.typ.sizedef {
-		a, _ = n.typ.zero()
-	} else {
-		a = reflect.MakeSlice(n.typ.TypeOf(), max, max)
-	}
-
+	typ := n.typ.frameType()
 	n.exec = func(f *frame) bltn {
+		var a reflect.Value
+		if n.typ.sizedef {
+			a, _ = n.typ.zero()
+		} else {
+			a = reflect.MakeSlice(typ, max, max)
+		}
 		for i, v := range values {
 			a.Index(index[i]).Set(v(f))
 		}
@@ -1622,8 +1630,16 @@ func mapLit(n *node) {
 	for i, c := range child {
 		convertLiteralValue(c.child[0], n.typ.key.TypeOf())
 		convertLiteralValue(c.child[1], n.typ.val.TypeOf())
-		keys[i] = genValue(c.child[0])
-		values[i] = genValue(c.child[1])
+		if n.typ.key.cat == interfaceT {
+			keys[i] = genValueInterface(c.child[0])
+		} else {
+			keys[i] = genValue(c.child[0])
+		}
+		if n.typ.val.cat == interfaceT {
+			values[i] = genValueInterface(c.child[1])
+		} else {
+			values[i] = genValue(c.child[1])
+		}
 	}
 
 	n.exec = func(f *frame) bltn {

--- a/interp/type.go
+++ b/interp/type.go
@@ -844,7 +844,7 @@ func (t *itype) zero() (v reflect.Value, err error) {
 		v, err = t.val.zero()
 
 	case arrayT, ptrT, structT:
-		v = reflect.New(t.TypeOf()).Elem()
+		v = reflect.New(t.frameType()).Elem()
 
 	case valueT:
 		v = reflect.New(t.rtype).Elem()
@@ -1002,7 +1002,7 @@ func exportName(s string) string {
 	return "X" + s
 }
 
-var interf = reflect.TypeOf(new(interface{})).Elem()
+var interf = reflect.TypeOf((*interface{})(nil)).Elem()
 
 // RefType returns a reflect.Type representation from an interpereter type.
 // In simple cases, reflect types are directly mapped from the interpreter

